### PR TITLE
Refactor: Merge build_moe_ffn_from_probs function into build_moe_ffn

### LINKED ISSUE #14920: Refactor: Merge build_moe_ffn_from_probs function into build_moe_ffn
### Background

As part of the work in PR #14898, the function `build_moe_ffn_from_probs` was introduced to handle SmallThinker's unique architecture where the MoE router is positioned before the attention block. This has resulted in unfortunate code duplication with the existing `build_moe_ffn` function.

### The Task

As suggested in [this code review comment](https://github.com/ggml-org/llama.cpp/pull/14898#discussion_r2234139265), the proposed solution is to merge the logic of `build_moe_ffn_from_probs` into the main `build_moe_ffn` function. This can be achieved by:

1.  Modifying `build_moe_ffn` to accept an optional `ggml_tensor *probs` parameter, which defaults to `nullptr`.
2.  Using this parameter as a "toggle":
    - If `probs` is provided, the function should use it directly and skip the internal logits/probs calculation.
    - If `probs` is `nullptr`, the function should behave as it currently does.
3.  Carefully handle the divergent logic paths inside the unified function, especially regarding weight normalization and activation functions.
4.  Once the merge is complete and verified, the now-redundant `build_moe_ffn_from_probs` function should be removed.

I plan to submit a pull request addressing this within the next 1-3 days.

### Context

- **Original PR:** #14898
- **Relevant Discussion:** https://github.com/ggml-org/llama.cpp/pull/14898#discussion_r2234139265

### PR DESCRIPTION
**Pull request by wdl339**
This PR resolves the code duplication discussed in Issue #14920 by merging the `build_moe_ffn_from_probs` function into the primary `build_moe_ffn` function.

### Description

Following the plan outlined in the issue, the unified `build_moe_ffn` function now accepts an optional `ggml_tensor *probs_in` parameter (defaulting to `nullptr`). This parameter acts as a "toggle" to control the logic path:

- If `probs_in` is provided, the function uses it directly, skipping the internal logits and probability calculations. This path is used by SmallThinker models where the MoE router is placed before the attention block.
- If `probs_in` is `nullptr`, the function maintains its original behavior, making it backward-compatible with existing MoE models.

Additionally, to fully support the SmallThinker model within this unified function, support for the ReLU activation function has been introduced.

The now-redundant `build_moe_ffn_from_probs` function has been removed. 

### Testing

I have performed basic inference tests on the following MoE models to ensure that the refactoring has not introduced any regressions:

- **SmallThinker-21BA3B-Instruct** (The new logic path)
- **Qwen3-30B-A3B** (The original logic path)

Both models appear to function correctly and produce coherent output.

Fixes #14920

### CODE REVIEW

**Comment by CISC** [comment]
Original logic path looks fine, but I'm having a hard time accepting that it's not regressing SmallThinker?

**Comment by wdl339** [comment]
> Original logic path looks fine, but I'm having a hard time accepting that it's not regressing SmallThinker?

That's an important question. The changes might look like a regression at first glance, and I appreciate you pushing for clarity on this. The new unified function is not a line-for-line copy of the old `build_moe_ffn_from_probs` logic. However,  it is functionally equivalent for SmallThinker.

### 1. The Core Architectural Difference (Why probs_in exists)
The primary architectural difference for SmallThinker is that its MoE router is positioned before the attention block. This means the router logits are calculated using the hidden state before the attention module. The calculation:
```
logits = build_lora_mm(gate_inp, cur); // [n_expert, n_tokens]
cb(logits, "ffn_moe_logits", il);
```
is moved into `llm_build_smallthinker`.

### 2. Simplifications in the Old `build_moe_ffn_from_probs`
The old `build_moe_ffn_from_probs` function was simpler. It omitted several checks, such as:
```
const bool weight_before_ffn = arch == LLM_ARCH_LLAMA4;
...
if (arch == LLM_ARCH_LLAMA4) {
  selection_probs = logits;
}
...
if (weight_before_ffn) {
  // repeat cur to [n_embd, n_expert_used, n_tokens]
  ggml_tensor * repeated = ggml_repeat_4d(ctx0, cur, n_embd, n_expert_used, n_tokens, 1);
  cur = ggml_mul(ctx0, repeated, weights);
  cb(cur, "ffn_moe_weighted", il);
}
```
and
```
if (scale_w) {
  weights = ggml_scale(ctx0, weights, w_scale);
  cb(weights, "ffn_moe_weights_scaled", il);
}
```
This was safe because those conditions would always be false for SmallThinker. In the new unified function, these checks remain for other models, but they correctly do not execute for SmallThinker.

### 3. The softmax/sigmoid Placement
I think this is the most confusing part, and where the apparent difference lies. Here's the breakdown:

Expert selection is unchanged. Since softmax and sigmoid are monotonic functions, applying them before `ggml_top_k` (as `build_moe_ffn`  does) or not (as `build_moe_ffn_from_probs`  did) does not change which experts are selected. The `ggml_top_k` and subsequent `ggml_get_rows` will yield the same selected_experts in both cases.

The old `build_moe_ffn_from_probs` applied softmax directly to the selected weights. Since softmax inherently normalizes its inputs to sum to 1, this allowed it to skip the explicit normalization block (`ggml_sum_rows` followed by `ggml_div`):
```
ggml_tensor * weights_sum = ggml_sum_rows(ctx0, weights); // [1, n_tokens]
cb(weights_sum, "ffn_moe_weights_sum", il);

weights = ggml_div(ctx0, weights, weights_sum); // [n_expert_used, n_tokens]
cb(weights, "ffn_moe_weights_norm", il);
```

When merging the functions, I made a decision to not carry over this specific, somewhat confusing optimization, to avoid complex if/else branching around the weight normalization logic.

I hope this explanation clarifies why the new implementation is functionally equivalent for SmallThinker. Please let me know if this makes sense, or if you spot any flaw in my reasoning.

**Comment by CISC** [comment]
Yes, `#3` was the sticking point, thank you for the very thorough explanation! :)

**Review by CISC — APPROVED** [review]
